go-controller/pkg/ovn/controller/services/services_controller.go

```diff
@@ -227,6 +235,128 @@ func (c *Controller) handleErr(err error, key interface{}) {
 	utilruntime.HandleError(err)
 }
 
+// initTopLevelCache will take load balancer data currently applied in OVN and populate the cache.
+// An important caveat here is that no effort is made towards populating some details of ovnlb.LB here.
+// That is because such work will be performed in syncService, so all that is needed here is the ability
+// to distinguish what is present in ovn database and this 'dirty' initial value.
+func (c *Controller) initTopLevelCache() error {
+	c.alreadyAppliedLock.Lock()
+	defer c.alreadyAppliedLock.Unlock()
+
+	// first, list all load balancers and their respective services
+	services, lbs, err := getServiceLBs(c.nbClient)
+	if err != nil {
+		return fmt.Errorf("failed to load balancers: %w", err)
+	}
+
+	c.alreadyApplied = make(map[string][]ovnlb.LB, len(services))
+
+	for _, lb := range lbs {
+		service := lb.ExternalIDs["k8s.ovn.org/owner"]
+		c.alreadyApplied[service] = append(c.alreadyApplied[service], *lb)
+	}
+
+	klog.Infof("Controller cache of %d load balancers initialized for %d services",
+		len(lbs), len(c.alreadyApplied))
+
+	return nil
+}
+
+// getServiceLBs lists all load balancers related to service from nbdb
+func getServiceLBs(nbClient libovsdbclient.Client) (sets.String, []*ovnlb.LB, error) {
+	lbs, err := libovsdbops.ListLoadBalancers(nbClient)
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not list load_balancer: %w", err)
+	}
+
+	services := sets.NewString()
+	out := make([]*ovnlb.LB, 0, len(lbs))
+	for _, lb := range lbs {
+
+		// Skip load balancers unrelated to service, or w/out an owner (aka namespace+name)
+		if lb.ExternalIDs["k8s.ovn.org/kind"] != "Service" {
+			continue
+		}
+
+		if service, ok := lb.ExternalIDs["k8s.ovn.org/owner"]; ok {
+			services.Insert(service)
+		} else {
+			continue
+		}
+
+		ps := func(item *nbdb.LogicalSwitch) bool {
+			for _, lbuuid := range item.LoadBalancer {
+				if lbuuid == lb.UUID {
+					return true
+				}
+			}
+			return false
+		}
+		switches, err := libovsdbops.FindLogicalSwitchesWithPredicate(nbClient, ps)
```

Comment:
I know this whole init will only be called once, but what do you think if we create a map of objects here like map[lb.UUID] =  LB without filling backrefs, and then in another loop after this one fetch all existing switches, routers and groups, and fill backrefs for already created LBs?

Reply:
Sure thing! It will definitely make it faster, I think.

Reply:
done.